File: packages/db/src/queries/conversations.ts
import { randomUUID } from "node:crypto";
import type { DB } from "../database.js";

export interface Conversation {
  conv_id: string;
  source: string;
  created_at: number;
  last_active: number;
  project_dir: string | null;
}

export function createConversation(db: DB, projectDir: string): Conversation {
  const now = Date.now();
  const conv: Conversation = {
    conv_id: randomUUID(),
    source: "local",
    created_at: now,
    last_active: now,
    project_dir: projectDir,
  };
  db.prepare(
    "INSERT INTO conversations (conv_id, source, created_at, last_active, project_dir) VALUES (?, ?, ?, ?, ?)",
  ).run(conv.conv_id, conv.source, conv.created_at, conv.last_active, conv.project_dir);
  return conv;
}

export function getConversation(db: DB, convId: string): Conversation | undefined {
  return db
    .prepare("SELECT * FROM conversations WHERE conv_id = ?")
    .get(convId) as Conversation | undefined;
}

export function getLatestConversation(db: DB): Conversation | undefined {
  return db
    .prepare("SELECT * FROM conversations ORDER BY last_active DESC LIMIT 1")
    .get() as Conversation | undefined;
}

export function touchConversation(db: DB, convId: string): void {
  db.prepare("UPDATE conversations SET last_active = ? WHERE conv_id = ?").run(
    Date.now(),
    convId,
  );
}

File: crates/oni-db/src/conversations.rs
use crate::schema::Database;
use oni_core::error::{Result, WrapErr};

pub struct StoredMessage {
    pub msg_id: String,
    pub role: String,
    pub content: String,
    pub tokens: i64,
    pub timestamp: String,
}

pub struct ConversationInfo {
    pub conv_id: String,
    pub source: String,
    pub created_at: String,
    pub last_active: String,
    pub project_dir: Option<String>,
}

impl Database {
    pub fn create_conversation(&self, project_dir: &str) -> Result<String> {
        let id = uuid::Uuid::new_v4().to_string();
        self.conn()
            .execute(
                "INSERT INTO conversations (conv_id, project_dir) VALUES (?1, ?2)",
                rusqlite::params![id, project_dir],
            )
            .wrap_err("Failed to create conversation")?;
        Ok(id)
    }

    pub fn add_message(&self, conv_id: &str, role: &str, content: &str) -> Result<String> {
        let id = uuid::Uuid::new_v4().to_string();
        let tokens = (content.len() / 4) as i64;
        self.conn()
            .execute(
                "INSERT INTO messages (msg_id, conv_id, role, content, tokens) VALUES (?1, ?2, ?3, ?4, ?5)",
                rusqlite::params![id, conv_id, role, content, tokens],
            )
            .wrap_err("Failed to add message")?;

        // Update last_active
        self.conn()
            .execute(
                "UPDATE conversations SET last_active = datetime('now') WHERE conv_id = ?1",
                [conv_id],
            )
            .wrap_err("Failed to update conversation")?;

        Ok(id)
    }

    pub fn get_messages(&self, conv_id: &str) -> Result<Vec<StoredMessage>> {
        let mut stmt = self
            .conn()
            .prepare(
                "SELECT msg_id, role, content, tokens, timestamp FROM messages WHERE conv_id = ?1 ORDER BY timestamp",
            )
            .wrap_err("Failed to prepare message query")?;
        let rows = stmt
            .query_map([conv_id], |row| {
                Ok(StoredMessage {
                    msg_id: row.get(0)?,
                    role: row.get(1)?,
                    content: row.get(2)?,
                    tokens: row.get(3)?,
                    timestamp: row.get(4)?,
                })
            })
            .wrap_err("Failed to query messages")?;
        Ok(rows.filter_map(|r| r.ok()).collect())
    }

    pub fn list_conversations(&self) -> Result<Vec<ConversationInfo>> {
        let mut stmt = self
            .conn()
            .prepare(
                "SELECT conv_id, source, created_at, last_active, project_dir FROM conversations ORDER BY last_active DESC",
            )
            .wrap_err("Failed to prepare conversation query")?;
        let rows = stmt
            .query_map([], |row| {
                Ok(ConversationInfo {
                    conv_id: row.get(0)?,
                    source: row.get(1)?,
                    created_at: row.get(2)?,
                    last_active: row.get(3)?,
                    project_dir: row.get(4)?,
                })
            })
            .wrap_err("Failed to query conversations")?;
        Ok(rows.filter_map(|r| r.ok()).collect())
    }
}
